### PR TITLE
feature/PT-359 osm export

### PR DESCRIPTION
**Pull request by wryczko-revolve**

### CODE REVIEW

File: OsmIntegrator/ApiModels/OsmExport/OsmChangeOutput.cs [hunk 1]
```diff
@@ -1,5 +1,7 @@
+using System.Collections.Generic;
+
 public class OsmChangeOutput
 {
-  public string Comment { get; set; }
-  public string OsmChangeFileContent { get; set; }
+  public string Changes { get; set; }
+  public IReadOnlyDictionary<string, string> Tags { get; set; }
```
Comment (luktar) [inline OsmIntegrator/ApiModels/OsmExport/OsmChangeOutput.cs:6]: I don't understand this change.

File: OsmIntegrator/Controllers/TileController.cs [hunk 2]
```diff
@@ -205,22 +205,31 @@ public async Task<ActionResult<Report>> UpdateStops(string id)
 
     [HttpGet("{id}")]
     [Authorize(Roles = UserRoles.EDITOR + "," + UserRoles.SUPERVISOR + "," + UserRoles.ADMIN + "," + UserRoles.COORDINATOR)]
-    public async Task<ActionResult<bool>> ContainsChanges(string id)
+    public async Task<ActionResult<bool>> ContainsChanges(Guid id)
     {
       DbTile tile = await GetTileAsync(id);
 
+      var minExportDelayMins = byte.Parse(_configuration["OsmExportMinDelayMins"]);
```
Comment (luktar) [inline OsmIntegrator/Controllers/TileController.cs:212]: 👍

File: OsmIntegrator/Controllers/TileController.cs [hunk 3]
```diff
@@ -205,22 +205,31 @@ public async Task<ActionResult<Report>> UpdateStops(string id)
...
       return Ok(containsChanges);
     }
 
-    private async Task<DbTile> GetTileAsync(string tileId)
+    private async Task<DbTile> GetTileAsync(Guid tileId)
     {
       DbTile currentTile = await _dbContext.Tiles
         .Include(tile => tile.Stops)
-        .SingleOrDefaultAsync(x => x.Id == Guid.Parse(tileId));
+        .Include(tile => tile.ExportReports)
```
Comment (luktar) [inline OsmIntegrator/Controllers/TileController.cs:231]: Does this method call always need ExportReports reference?

File: OsmIntegrator/Database/Models/DbTile.cs [hunk 4]
```diff
@@ -52,8 +55,63 @@ public class DbTile
...
     public List<DbConversation> Conversations { get; set; }
 
-    public List<DbChangeReport> ChangeReports { get; set; }
+    public List<DbChangeReport> ChangeReports { get; set; } = new();
 
+    public List<DbTileExportReport> ExportReports { get; set; } = new();
+
+    public DateTime? LastExportAt => ExportReports.Any() ? ExportReports.Select(r => r.CreatedAt).Max() : null;
+
+    public string GetUnexportedChanges(IStringLocalizer _localizer)
```
Comment (luktar) [inline OsmIntegrator/Database/Models/DbTile.cs:64]: I don't think it's a good place for this method ;).

File: OsmIntegrator/Database/Models/DbTileExportReport.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+using System.ComponentModel.DataAnnotations.Schema;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Identity;
+using OsmIntegrator.Database.Models;
+using OsmIntegrator.Database.Models.JsonFields;
+
+namespace OsmIntegrator.Database.Models
+{
+  [Table("TileExportReport")]
```
Comment (luktar) [inline OsmIntegrator/Database/Models/DbTileExportReport.cs]: I have the rule in the project that the DB table name ends with `s`. So if we have a class `DbConnection` the DB table name is called `Connections`.

File: OsmIntegrator/Services/OsmExporter.cs [hunk 6]
```diff
@@ -105,5 +108,13 @@ public string GetComment(long x, long y, byte zoom)
       sb.Append("Wiki: https://wiki.openstreetmap.org/w/index.php?title=Automated_edits/luktar/OsmIntegrator_-_fixing_stop_signs_for_blind");
       return sb.ToString();
     }
+
+    public IReadOnlyDictionary<string, string> GetTags(string comment) => new Dictionary<string, string> {
```
Comment (luktar) [inline OsmIntegrator/Services/OsmExporter.cs:112]: 👍

File: OsmIntegrator/appsettings.json [hunk 7]
```diff
@@ -18,6 +18,7 @@
   },
   "AllowedHosts": "*",
   "OsmCronInterval": "0 0 2 * * *",
+  "OsmExportMinDelayMins": "5",
```
Comment (luktar) [inline OsmIntegrator/appsettings.json:21]: 👍

**Review by luktar — APPROVED** [review]

File: OsmIntegrator/ApiModels/OsmExport/OsmChangeOutput.cs (hunk 1 above)
Comment (wryczko-revolve) [reply OsmIntegrator/ApiModels/OsmExport/OsmChangeOutput.cs:6]: discussed

File: OsmIntegrator/Database/Models/DbTile.cs (hunk 4 above)
Comment (wryczko-revolve) [reply OsmIntegrator/Database/Models/DbTile.cs:64]: discussed

File: OsmIntegrator/ApiModels/OsmExport/OsmChangeOutput.cs (hunk 1 above)
Comment (wryczko-revolve) [reply OsmIntegrator/ApiModels/OsmExport/OsmChangeOutput.cs:6]: discussed

File: OsmIntegrator/Database/Models/DbTile.cs (hunk 4 above)
Comment (wryczko-revolve) [reply OsmIntegrator/Database/Models/DbTile.cs:64]: discussed

File: OsmIntegrator/Database/Models/DbTileExportReport.cs (hunk 5 above)
Comment (wryczko-revolve) [reply OsmIntegrator/Database/Models/DbTileExportReport.cs]: renamed

**Pull request merged**